### PR TITLE
Bot code improvement part 3 - Removes bot core.

### LINKED ISSUE #63259: Simple robot control locks can be disabled without the proper ID access
Reporting client version: 514.1556

## Round ID:
[174195](https://scrubby.melonmesa.com/round/174195)

## Testmerges:
- [Families Bugfixes + Tweaks: Even Distribution of Leaders, Fixes HoS-Family-Leaders, Adds Admin Tool to make Leaders](https://www.github.com/tgstation/tgstation/pull/63115)

## Reproduction:
Find a simple bot(Cleanbot, secbot, etc).
Click on it and open the settings panel.
Disable the controls lock without the proper ID.

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

This deletes bot_core entirely, instead the access required is on the bot itself

My previous refactor broke bot access cover requirements, and that bug is what inspired me to make this refactor.

## Why It's Good For The Game

This is a stupid thing to have on the bot, I have no idea why it is a thing, and has only ever confused me since the very start. I find this functions better, and makes the code more easy to understand.

Closes https://github.com/tgstation/tgstation/issues/63259

## Changelog

:cl:
fix: Bot covers now respect access again.
/:cl:

### CODE REVIEW

File: code/modules/modular_computers/file_system/programs/robocontrol.dm [hunk 1]
```diff
@@ -35,20 +35,37 @@
...
-	for(var/B in GLOB.bots_list)
-		var/mob/living/simple_animal/bot/Bot = B
-		if(!(Bot.bot_mode_flags & BOT_MODE_ON) || Bot.z != zlevel || !(Bot.bot_mode_flags & BOT_MODE_REMOTE_ENABLED)) //Only non-emagged bots on the same Z-level are detected!
+	for(var/mob/living/simple_animal/bot/Bot as anything in GLOB.bots_list)
+		if(Bot.z != zlevel || !(Bot.bot_mode_flags & BOT_MODE_REMOTE_ENABLED)) //Only non-emagged bots on the same Z-level are detected!
 			continue
 		else if(computer) //Also, the inserted ID must have access to the bot type
 			var/obj/item/card/id/id_card = card_slot ? card_slot.stored_card : null
-			if(!id_card && !Bot.bot_core.allowed(current_user))
+			if(!id_card && !Bot.check_access(current_user))
```
Comment (AnturK) [inline code/modules/modular_computers/file_system/programs/robocontrol.dm:43]: You're not passing in the card at all here.

**Review by AnturK — COMMENTED** [review]
Correct there's no reason for this to exist.

File: code/modules/modular_computers/file_system/programs/robocontrol.dm (hunk 1 above)
Comment (JohnFulpWillard) [reply code/modules/modular_computers/file_system/programs/robocontrol.dm]: it wasnt needed anymore so I just removed that now

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 2]
```diff
@@ -229,9 +227,24 @@
...
+
+	var/obj/item/card/id/id_card = user.get_idcard(TRUE)
+	if(!id_card || !id_card.access)
+		return FALSE
+	id_card = id_card.GetID()
+
+	for(var/requested_access in maints_access_required)
+		if(requested_access in id_card.access)
+			return TRUE
+	return FALSE
```
Comment (Watermelon914) [inline code/modules/mob/living/simple_animal/bot/bot.dm:246]: feels a bit similar to object's check_access_list, but I suppose it is distinct enough

File: code/modules/modular_computers/file_system/programs/robocontrol.dm [hunk 3]
```diff
@@ -35,20 +35,34 @@
 	botcount = 0
 	current_user = user
 
-	for(var/B in GLOB.bots_list)
-		var/mob/living/simple_animal/bot/Bot = B
-		if(!(Bot.bot_mode_flags & BOT_MODE_ON) || Bot.z != zlevel || !(Bot.bot_mode_flags & BOT_MODE_REMOTE_ENABLED)) //Only non-emagged bots on the same Z-level are detected!
+	for(var/mob/living/simple_animal/bot/Bot as anything in GLOB.bots_list)
```
Comment (Watermelon914) [inline code/modules/modular_computers/file_system/programs/robocontrol.dm:38]:
```suggestion
	for(var/mob/living/simple_animal/bot/bot as anything in GLOB.bots_list)
```

Comment (Watermelon914) [reply code/modules/modular_computers/file_system/programs/robocontrol.dm:38]: also tweak the variable name in the for proc as well

File: code/modules/modular_computers/file_system/programs/robocontrol.dm [hunk 4]
```diff
@@ -35,20 +35,34 @@
...
+			"model" = Bot.bot_type,
+			"locat" = get_area(Bot),
+			"bot_ref" = REF(Bot),
+			"mule_check" = FALSE,
+		)
 		if(Bot.bot_type == MULE_BOT)
 			var/mob/living/simple_animal/bot/mulebot/MULE = Bot
-			mulelist += list(list("name" = MULE.name, "dest" = MULE.destination, "power" = MULE.cell ? MULE.cell.percent() : 0, "home" = MULE.home_destination, "autoReturn" = MULE.auto_return, "autoPickup" = MULE.auto_pickup, "reportDelivery" = MULE.report_delivery, "mule_ref" = REF(MULE)))
+			mulelist += list(
+				list(
```
Comment (Watermelon914) [inline code/modules/modular_computers/file_system/programs/robocontrol.dm:55]:
```suggestion
			mulelist += list(list(
```

File: code/modules/modular_computers/file_system/programs/robocontrol.dm [hunk 5]
```diff
@@ -35,20 +35,34 @@
...
+					"name" = MULE.name,
+					"dest" = MULE.destination,
+					"power" = MULE.cell ? MULE.cell.percent() : 0,
+					"home" = MULE.home_destination,
+					"autoReturn" = MULE.auto_return,
+					"autoPickup" = MULE.auto_pickup,
+					"reportDelivery" = MULE.report_delivery,
+					"mule_ref" = REF(MULE),
+				),
+			)
```
Comment (Watermelon914) [inline code/modules/modular_computers/file_system/programs/robocontrol.dm:65]:
```suggestion
		))
```

File: code/modules/modular_computers/file_system/programs/robocontrol.dm (hunk 4 above)
Comment (Watermelon914) [reply code/modules/modular_computers/file_system/programs/robocontrol.dm:55]: this'll just shorten the code, makes it take less space and is easier to read

File: code/modules/modular_computers/file_system/programs/robocontrol.dm (hunk 3 above)
Comment (JohnFulpWillard) [reply code/modules/modular_computers/file_system/programs/robocontrol.dm]: all done

File: code/modules/modular_computers/file_system/programs/robocontrol.dm [hunk 6]
```diff
@@ -35,20 +35,31 @@
...
+		var/list/newbot = list(
+			"name" = simple_bot.name,
+			"mode" = simple_bot.get_mode_ui(),
+			"model" = simple_bot.bot_type,
+			"locat" = get_area(simple_bot),
+			"bot_ref" = REF(simple_bot),
+			"mule_check" = FALSE,
+		)
+		if(simple_bot.bot_type == MULE_BOT)
+			var/mob/living/simple_animal/bot/mulebot/MULE = simple_bot
```
Comment (Watermelon914) [inline code/modules/modular_computers/file_system/programs/robocontrol.dm:52]: same thing here ideally, keep variable names camel case

Comment (JohnFulpWillard) [reply code/modules/modular_computers/file_system/programs/robocontrol.dm]: i went ahead and did more even further down just to be safe

**Pull request merged**